validation/topsort.go

```diff
@@ -0,0 +1,107 @@
+//copy from https://github.com/philopon/go-toposort
+package validation
+
+type Graph struct {
+	nodes   []string
+	outputs map[string]map[string]int
+	inputs  map[string]int
+}
+
+func NewGraph() *Graph {
+	return &Graph{
+		nodes:   make([]string, 0),
+		inputs:  make(map[string]int),
+		outputs: make(map[string]map[string]int),
+	}
+}
+
+func (g *Graph) AddNode(name string) bool {
+	g.nodes = append(g.nodes, name)
+
+	if _, ok := g.outputs[name]; ok {
+		return false
+	}
+	g.outputs[name] = make(map[string]int)
+	g.inputs[name] = 0
+	return true
+}
+
+func (g *Graph) AddNodes(names ...string) bool {
+	for _, name := range names {
+		if ok := g.AddNode(name); !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func (g *Graph) AddEdge(from, to string) bool {
+	m, ok := g.outputs[from]
+	if !ok {
+		return false
+	}
+
+	m[to] = len(m) + 1
+	g.inputs[to]++
+
+	return true
+}
+
+func (g *Graph) IsNodePresent(name string) bool {
+	_, ok := g.outputs[name]
+	return ok
+}
+
+func (g *Graph) unsafeRemoveEdge(from, to string) {
+	delete(g.outputs[from], to)
+	g.inputs[to]--
+}
+
+func (g *Graph) RemoveEdge(from, to string) bool {
+	if _, ok := g.outputs[from]; !ok {
+		return false
+	}
+	g.unsafeRemoveEdge(from, to)
+	return true
+}
+
+func (g *Graph) Toposort() ([]string, bool) {
+	L := make([]string, 0, len(g.nodes))
+	S := make([]string, 0, len(g.nodes))
+
+	for _, n := range g.nodes {
+		if g.inputs[n] == 0 {
+			S = append(S, n)
+		}
+	}
+
+	for len(S) > 0 {
+		var n string
+		n, S = S[0], S[1:]
+		L = append(L, n)
+
+		ms := make([]string, len(g.outputs[n]))
+		for m, i := range g.outputs[n] {
+			ms[i-1] = m
+		}
+
+		for _, m := range ms {
+			g.unsafeRemoveEdge(n, m)
+
+			if g.inputs[m] == 0 {
+				S = append(S, m)
+			}
+		}
+	}
+
+	N := 0
+	for _, v := range g.inputs {
+		N += v
+	}
+
+	if N > 0 {
+		return L, false
+	}
+
+	return L, true
+}
```

Comment:
My initial assumption was that you want to use topsort in order to produce better error messages when a circular dependency exists. This would rely entirely on `L` here but I don't think that this yields useful information in case of a circle but only the topologically sorted order if there is no such circle.

For instance given this example:
![khan1](https://cloud.githubusercontent.com/assets/733004/26035005/f9fae854-38c5-11e7-8d46-efa7f498b930.png)

This would start with `S= {1}` then move node `1` from `S` to `L` and remove the three edges `[1,2]`, `[1,3]` and `[1,4]`. But it would only add `2` to `S` but neither `3` or `4` (since they still contain other incoming edges). Then in the next iteration `2` is moved from `S` to `L` and the algorithm stops with `L={1,2}` because there are no more outgoing edges from `2`.

I am no longer sure why you want to replace the currently implemented depth-first search with topsort. Can you explain what problem you actually want to solve with this?